lib/6to5/traverse/index.js

```diff
@@ -50,104 +57,139 @@ TraversalContext.prototype.maybeReplace = function (result, obj, key, node) {
   }
 
   if (isArray) {
-    this.flatten();
+    return true;
+  }
+}
+
+TraversalContext.prototype.enterNode = function (obj, key, node, enter, parent, scope, state) {
+  var result = enter(node, parent, scope, this, state);
+  var flatten = false;
+
+  if (result) {
+    flatten = replaceNode(obj, key, node, result);
+    node = result;
+
+    if (flatten) {
+      this.didFlatten = true;
+    }
+  }
+
+  if (this.didRemove) {
+    obj[key] = null;
+    this.didFlatten = true;
+  }
+
+  return node;
+};
+
+TraversalContext.prototype.exitNode = function (obj, key, node, exit, parent, scope, state) {
+  var result = exit(node, parent, scope, this, state);
+  var flatten = false;
+
+  if (result) {
+    flatten = replaceNode(obj, key, node, result);
+    node = result;
+
+    if (flatten) {
+      this.didFlatten = true;
+    }
   }
 
   return node;
 };
 
-TraversalContext.prototype.visit = function (obj, key, opts, scope, parent) {
+TraversalContext.prototype.visitNode = function (obj, key, opts, scope, parent, state) {
+  this.reset();
+
   var node = obj[key];
-  if (!node) return;
 
   // type is blacklisted
-  if (opts.blacklist && opts.blacklist.indexOf(node.type) > -1) return;
+  if (opts.blacklist && opts.blacklist.indexOf(node.type) > -1)
```

Comment:
Ya, a linear scan has some great benefits for small N. Assuming the stars align.

Im curious what the cliff is, once i'm off battery power I will investigate that further.